Fix: Infinite loop caused by calling $order->save_meta_data() during `woocommerce_update_order` hook

prettyboymp opened the pull request:
### Changes proposed in this Pull Request:

This modifies Abstract_WC_ORder_Data_Store_CPT->clear_caches() to also remove in-memory OrderCache entry for the given order.

There is common pattern in extensions to do the following:

```
add_action( 'woocommerce_update_order', function($order_id, $order) {
	$order = wc_get_order( $order_id );
	$order->update_meta_data( 'some_meta_key', 'some_meta_value' );
	$order->save_meta_data();
}, 10, 2 );
```

When this is run on an order for the first time, the meta data entry is added and saved, but ends up in an infinite loop, because of the following scenario:

1. `$order = wc_get_order( $order_id );` retrieves the previously instantiated Order instance from in-memory OrderCache,
2. `$order->save_meta_data();` Saves the changes back to the DB, but then again runs the `woocommerce_update_order` hook before the `OrderCache` entry is moved.
3. Step 1 repeats again because the in-memory copy doesn't yet see the meta entry.


<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOPLUG-4910 / #59455  .

(For Bug Fixes) Bug introduced in PR #46023 .


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

To test that it fixes the issue with the infinite loop:

**Requires object caching**

1. Enable HPOS and  `Enable order data caching in the datastore` under WooCommerce -> Settings -> Advanced -> Features.
2. Create an order through any means or identify an existing order.
3. Add the following code snippet somewhere it will execute:
```
add_action( 'woocommerce_update_order', function($order_id, $order) {
	$order = wc_get_order( $order_id );
	$order->update_meta_data( 'some_meta_key', 'some_meta_value' );
	$order->save_meta_data();
}, 10, 2 );
```
4. Go to edit the order through the admin and save it.  - Note, this only works once as the original issue only occurred the first time new meta was added during the `woocommerce_update_order` hook.  So you will need to either change the meta key you're testing with or use a new order each time.

If the problem persists, doing the above will trigger an infinite loop.  Otherwise, it should save normally.

General testing - 
Do some general testing around orders.
* Make sure you can place, create, edit, and delete orders and there are no noticeable issues where incorrect information is showing.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

prettyboymp committed:
Remove in-memory OrderCache entry when Abstract_WC_ORder_Data_Store_CPT->clear_caches() is called.

prettyboymp committed:
Add changelog

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-4910/clear-order-cache-after-meta-save

prettyboymp committed:
Clear caches after a refund update.

kalessil commented on plugins/woocommerce/includes/abstracts/abstract-wc-order.php:
```diff
@@ -226,11 +226,6 @@ public function save() {
 
 			$this->save_items();
 
-			if ( OrderUtil::orders_cache_usage_is_enabled() ) {
```
It appears that we have some unused imports in this file with this change.

kalessil commented on plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php:
```diff
@@ -503,6 +503,10 @@ protected function clear_caches( &$order ) {
 		clean_post_cache( $order->get_id() );
 		wc_delete_shop_order_transients( $order );
 		wp_cache_delete( 'order-items-' . $order->get_id(), 'orders' );
+		if ( \Automattic\WooCommerce\Utilities\OrderUtil::orders_cache_usage_is_enabled() ) {
```
For consistency, we should use proper imports here instead of FQNs.

kalessil commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableRefundDataStore.php:
```diff
@@ -143,6 +143,7 @@ public function create( &$refund ) {
 	 */
 	public function update( &$refund ) {
 		$this->persist_updates( $refund );
+		$this->clear_caches( $refund );
```
Not sure if this call is necessary: `persist_updates` will clear the cache due to `$backfill = true` fallback. Can you please verify?

prettyboymp replied:
Good catch on this.  I added it because the test `OrdersTableDataStoreTests::test_error_when_setting_order_property_is_captured_and_logged()` was failing, but after your prompt and relooking at it, I realized it was technically an issue with the test. If something makes a direct update to the DB, it is expected that that code will need to clear that order object from `OrderCache`.  The backfilling was re-priming the cache which caused the test to fail.

I did at least find an fix another issue with refunds, which was that it wasn't applying the changes to the refund after an update.

kalessil replied:
Nice! I'm reviewing and manually testing this PR.

kalessil reviewed (commented):
LGTM to me overall. @prettyboymp can you please take a look into my comments before I start manual testing?

prettyboymp committed:
Switch to using imports
Remove the call to clear_cache  in OrdersTableRefundDataStore, it was only the test that needed updated.

kalessil reviewed (approved):
LGTM!

- Testing scenario :heavy_check_mark: 
- Keeping the snippet in place: create an order and refund the order :heavy_check_mark:

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-4910/clear-order-cache-after-meta-save

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-4910/clear-order-cache-after-meta-save

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-4910/clear-order-cache-after-meta-save
